### PR TITLE
Add Wishlists LIST and QUERY API

### PR DESCRIPTION
**Pull request by sqz0914**
1. Add wishlist list and query feature
2. Add corresponding test cases
3. Add more data model test cases
The current coverage does not change:
Screen Shot 2020-10-17 at 4 27 38 AM
4. Fix some pep8 format issues and now there is only one left which seems no need to make the change:
Screen Shot 2020-10-17 at 4 28 00 AM
5. Modify the test case for 405 METHOD_NOT_ALLOWED using mock so as to make it more related to our services

### CODE REVIEW

**Review by xiafanzeng — DISMISSED** [review]
Cause duplication of work. The work here in LIST covers the task of QUERY (find by user id and wishlist name), will close the QUERY story.

**Comment by abhinavGupta16** [comment]
> Cause duplication of work. The work here in LIST covers the task of QUERY (find by user id and wishlist name), will close the QUERY story.

Moved back to QA, as PR is not yet merged

**Review by abhinavGupta16 — CHANGES_REQUESTED** [review]
Needs test cases for unhappy paths - 404, 415, 400 etc.

**Comment by sqz0914** [comment]
> Needs test cases for unhappy paths - 404, 415, 400 etc.

Do you refer to QUERY or both?

File: tests/test_models.py [hunk 1]
```diff
@@ -43,10 +44,132 @@ def tearDown(self):
...
+
+    def test_update_wishlist(self):
+        """ Update a wishlist """
+        wishlist = self._create_wishlist()
+        wishlist.create()
+        # Assert that it was assigned an id and shows up in the database
+        self.assertEqual(wishlist.id, 1)
+
+        # Fetch it back
+        wishlist = Wishlist.find_or_404(wishlist.id)
```
Comment (abhinavGupta16) [inline tests/test_models.py:112]: If this throws 404, the test fails with Exception and not Assertion. Same for rest.

File: tests/test_service.py [hunk 2]
```diff
@@ -62,7 +62,7 @@ def tearDown(self):
         db.session.remove()
         db.drop_all()
 
-    def _create_wishlists(self, count=1):
+    def _create_wishlists(self, count):
```
Comment (abhinavGupta16) [inline tests/test_service.py:66]: I think we need to have count = 1. This just sets the default value of count to 1. Makes the function easier to use.

File: tests/test_service.py [hunk 3]
```diff
@@ -232,7 +243,7 @@ def test_add_item_to_wishlist(self):
         # Check the data is correct
         resp_item = resp.get_json()
 
-        self.assertEqual(resp_item["wishlist_id"], new_item.id,
+        self.assertEqual(resp_item["wishlist_id"], new_item.wishlist_id,
```
Comment (abhinavGupta16) [inline tests/test_service.py:251]: Good catch

**Review by abhinavGupta16 — CHANGES_REQUESTED** [review]

**Comment by abhinavGupta16** [comment]
> > Needs test cases for unhappy paths - 404, 415, 400 etc.
> 
> Do you refer to QUERY or both?

Both. wherever applicable.

**Comment by abhinavGupta16** [comment]
Also check, flake8 --count --max-complexity=10 --statistics tests

**Comment by sqz0914** [comment]
> Also check, flake8 --count --max-complexity=10 --statistics tests

Do you read my description of the changes? Go up and see the 4th change. It is about fixing some format issues.

**Comment by sqz0914** [comment]
> > > Needs test cases for unhappy paths - 404, 415, 400 etc.
> > 
> > 
> > Do you refer to QUERY or both?
> 
> Both. wherever applicable.

For the NOT FOUND case, if the desired wishlists are not found, an empty wishlist will be returned. The list_wishlists() will still work and return a status code of 200. For UNSOPPORTED MEDIA TYPE, since both services do not put or post any json objects, the content type check is not needed. For BAD_REQUEST, could you give an example of such error? Since no data is put or post, I don't think it is necessary as well.

File: tests/test_models.py (hunk 1 above)
Comment (sqz0914) [reply tests/test_models.py:112]: I don't catch your point. The test is fine. Just download the code and check. Same for rest.

File: tests/test_service.py (hunk 2 above)
Comment (sqz0914) [reply tests/test_service.py:66]: How could we assume when developers want to use this helper function, they only need to create one wishlist by default? Also note that the naming of this function is _create_wishlist**s**, not _create_wishlist.

**Comment by abhinavGupta16** [comment]
> > > > Needs test cases for unhappy paths - 404, 415, 400 etc.
> > > 
> > > 
> > > Do you refer to QUERY or both?
> > 
> > 
> > Both. wherever applicable.
> 
> For the NOT FOUND case, if the desired wishlists are not found, an empty wishlist will be returned. The list_wishlists() will still work and return a status code of 200. For UNSOPPORTED MEDIA TYPE, since both services do not put or post any json objects, the content type check is not needed. For BAD_REQUEST, could you give an example of such error? Since no data is put or post, I don't think it is necessary as well.

Yep, if there if no POST or put, there will not be a 415, hence, where applicable. 
Bad request can be if the user_id is a string? 
404 - can be the ids or name are not found.

File: service/models.py [hunk 4]
```diff
@@ -226,3 +224,35 @@ def deserialize(self, data: dict):
...
+
+        :return: a collection of Wishlists with that name
+        :rtype: list
+
+        """
+        cls.logger.info("Processing name query for %s ...", name)
+        return cls.query.filter(cls.name == name)
+
+    @classmethod
+    def find_by_user_id(cls, user_id: str):
```
Comment (abhinavGupta16) [inline service/models.py]: user_id should be an integer.

**Review by abhinavGupta16 — CHANGES_REQUESTED** [review]

**Comment by sqz0914** [comment]
> > > > > Needs test cases for unhappy paths - 404, 415, 400 etc.
> > > > 
> > > > 
> > > > Do you refer to QUERY or both?
> > > 
> > > 
> > > Both. wherever applicable.
> > 
> > 
> > For the NOT FOUND case, if the desired wishlists are not found, an empty wishlist will be returned. The list_wishlists() will still work and return a status code of 200. For UNSOPPORTED MEDIA TYPE, since both services do not put or post any json objects, the content type check is not needed. For BAD_REQUEST, could you give an example of such error? Since no data is put or post, I don't think it is necessary as well.
> 
> Yep, if there if no POST or put, there will not be a 415, hence, where applicable.
> Bad request can be if the user_id is a string?
> 404 - can be the ids or name are not found.

If the ids or names are not found, an empty list of wishlists will be returned, and the status code would be 200. You can use Postman to check. If the user_id is a string, it will be a 500 error.

**Comment by abhinavGupta16** [comment]
> > > > > > Needs test cases for unhappy paths - 404, 415, 400 etc.
> > > > > 
> > > > > 
> > > > > Do you refer to QUERY or both?
> > > > 
> > > > 
> > > > Both. wherever applicable.
> > > 
> > > 
> > > For the NOT FOUND case, if the desired wishlists are not found, an empty wishlist will be returned. The list_wishlists() will still work and return a status code of 200. For UNSOPPORTED MEDIA TYPE, since both services do not put or post any json objects, the content type check is not needed. For BAD_REQUEST, could you give an example of such error? Since no data is put or post, I don't think it is necessary as well.
> > 
> > 
> > Yep, if there if no POST or put, there will not be a 415, hence, where applicable.
> > Bad request can be if the user_id is a string?
> > 404 - can be the ids or name are not found.
> 
> If the ids or names are not found, an empty list of wishlists will be returned, and the status code would be 200. You can use Postman to check. If the user_id is a string, it will be a 500 error.

if the name is not found, we get a 404, but if the user_id is not a number, we shouldn't query the database as the request itself is invalid.

**Comment by sqz0914** [comment]
> > > > > > > Needs test cases for unhappy paths - 404, 415, 400 etc.
> > > > > > 
> > > > > > 
> > > > > > Do you refer to QUERY or both?
> > > > > 
> > > > > 
> > > > > Both. wherever applicable.
> > > > 
> > > > 
> > > > For the NOT FOUND case, if the desired wishlists are not found, an empty wishlist will be returned. The list_wishlists() will still work and return a status code of 200. For UNSOPPORTED MEDIA TYPE, since both services do not put or post any json objects, the content type check is not needed. For BAD_REQUEST, could you give an example of such error? Since no data is put or post, I don't think it is necessary as well.
> > > 
> > > 
> > > Yep, if there if no POST or put, there will not be a 415, hence, where applicable.
> > > Bad request can be if the user_id is a string?
> > > 404 - can be the ids or name are not found.
> > 
> > 
> > If the ids or names are not found, an empty list of wishlists will be returned, and the status code would be 200. You can use Postman to check. If the user_id is a string, it will be a 500 error.
> 
> if the name is not found, we get a 404, but if the user_id is not a number, we shouldn't query the database as the request itself is invalid.

Screen Shot 2020-10-17 at 1 55 42 PM
Please check the name not found case with postman.

**Comment by abhinavGupta16** [comment]
> > > > > > > > Needs test cases for unhappy paths - 404, 415, 400 etc.
> > > > > > > 
> > > > > > > 
> > > > > > > Do you refer to QUERY or both?
> > > > > > 
> > > > > > 
> > > > > > Both. wherever applicable.
> > > > > 
> > > > > 
> > > > > For the NOT FOUND case, if the desired wishlists are not found, an empty wishlist will be returned. The list_wishlists() will still work and return a status code of 200. For UNSOPPORTED MEDIA TYPE, since both services do not put or post any json objects, the content type check is not needed. For BAD_REQUEST, could you give an example of such error? Since no data is put or post, I don't think it is necessary as well.
> > > > 
> > > > 
> > > > Yep, if there if no POST or put, there will not be a 415, hence, where applicable.
> > > > Bad request can be if the user_id is a string?
> > > > 404 - can be the ids or name are not found.
> > > 
> > > 
> > > If the ids or names are not found, an empty list of wishlists will be returned, and the status code would be 200. You can use Postman to check. If the user_id is a string, it will be a 500 error.
> > 
> > 
> > if the name is not found, we get a 404, but if the user_id is not a number, we shouldn't query the database as the request itself is invalid.
> 
> Screen Shot 2020-10-17 at 1 55 42 PM
> 
> Please check the name not found case with postman.

Yes, it's a design call. We can either choose to give an empty list of a 404. However, user_id given should be tested for integer or bad_data with an error message like user_id needs to be an Integer.

File: service/models.py (hunk 4 above)
Comment (sqz0914) [reply service/models.py]: Has changed.

File: tests/test_models.py (hunk 1 above)
Comment (Jesse1204) [reply tests/test_models.py:112]: This test looks good to me. This should not throw 404. If it does, it means that our code has a problem. That's the purpose of unit tests.

**Comment by sqz0914** [comment]
> > > > > > > > > Needs test cases for unhappy paths - 404, 415, 400 etc.
> > > > > > > > 
> > > > > > > > 
> > > > > > > > Do you refer to QUERY or both?
> > > > > > > 
> > > > > > > 
> > > > > > > Both. wherever applicable.
> > > > > > 
> > > > > > 
> > > > > > For the NOT FOUND case, if the desired wishlists are not found, an empty wishlist will be returned. The list_wishlists() will still work and return a status code of 200. For UNSOPPORTED MEDIA TYPE, since both services do not put or post any json objects, the content type check is not needed. For BAD_REQUEST, could you give an example of such error? Since no data is put or post, I don't think it is necessary as well.
> > > > > 
> > > > > 
> > > > > Yep, if there if no POST or put, there will not be a 415, hence, where applicable.
> > > > > Bad request can be if the user_id is a string?
> > > > > 404 - can be the ids or name are not found.
> > > > 
> > > > 
> > > > If the ids or names are not found, an empty list of wishlists will be returned, and the status code would be 200. You can use Postman to check. If the user_id is a string, it will be a 500 error.
> > > 
> > > 
> > > if the name is not found, we get a 404, but if the user_id is not a number, we shouldn't query the database as the request itself is invalid.
> > 
> > 
> > Screen Shot 2020-10-17 at 1 55 42 PM
> > Please check the name not found case with postman.
> 
> Yes, it's a design call. We can either choose to give an empty list of a 404. However, user_id given should be tested for integer or bad_data with an error message like user_id needs to be an Integer.

I think it is no need to manually create this 404 error. I do add data type error check and the test case for that.

> > > > > > > > > Needs test cases for unhappy paths - 404, 415, 400 etc.
> > > > > > > > 
> > > > > > > > 
> > > > > > > > Do you refer to QUERY or both?
> > > > > > > 
> > > > > > > 
> > > > > > > Both. wherever applicable.
> > > > > > 
> > > > > > 
> > > > > > For the NOT FOUND case, if the desired wishlists are not found, an empty wishlist will be returned. The list_wishlists() will still work and return a status code of 200. For UNSOPPORTED MEDIA TYPE, since both services do not put or post any json objects, the content type check is not needed. For BAD_REQUEST, could you give an example of such error? Since no data is put or post, I don't think it is necessary as well.
> > > > > 
> > > > > 
> > > > > Yep, if there if no POST or put, there will not be a 415, hence, where applicable.
> > > > > Bad request can be if the user_id is a string?
> > > > > 404 - can be the ids or name are not found.
> > > > 
> > > > 
> > > > If the ids or names are not found, an empty list of wishlists will be returned, and the status code would be 200. You can use Postman to check. If the user_id is a string, it will be a 500 error.
> > > 
> > > 
> > > if the name is not found, we get a 404, but if the user_id is not a number, we shouldn't query the database as the request itself is invalid.
> > 
> > 
> > Screen Shot 2020-10-17 at 1 55 42 PM
> > Please check the name not found case with postman.
> 
> Yes, it's a design call. We can either choose to give an empty list of a 404. However, user_id given should be tested for integer or bad_data with an error message like user_id needs to be an Integer.

Just as discussed, no need to manually add this 404 error, an empty list is enough to tell the customer the search result is not existed. For the bad data, I add data type check of user_id and name, and add test case as well.

**Review by Jesse1204 — DISMISSED** [review]

**Comment by abhinavGupta16** [comment]
@sqz0914 - Looks good to me. Just need to fix the flake8 compliance.

flake8 --count --max-complexity=10 --statistics tests

image

**Review by abhinavGupta16 — CHANGES_REQUESTED** [review]
@sqz0914 - Looks good to me. Just need to fix the flake8 compliance.

flake8 --count --max-complexity=10 --statistics tests

image

**Review by abhinavGupta16 — APPROVED** [review]

**Review by Jesse1204 — APPROVED** [review]

**Pull request merged**